File: examples/frameworks/langchain/verify_free.py
"""Run LangChain MCPAdapter discovery and two fixed synthetic example tools only."""
import asyncio
import hashlib
import importlib.metadata
import json
import os
import platform
from datetime import datetime, timezone
from pathlib import Path

os.environ.update(OTEL_SDK_DISABLED="true", LANGCHAIN_TRACING_V2="false", LANGSMITH_TRACING="false", DO_NOT_TRACK="1", ANONYMIZED_TELEMETRY="false")
for key in tuple(os.environ):
    if key.endswith("API_KEY") or key.startswith(("LANGSMITH_API", "LANGCHAIN_API")):
        os.environ.pop(key, None)

import httpx
import httpx2

ROOT = Path(__file__).resolve().parent
ENDPOINT = "https://agents.getardaro.com/mcp"
FREE = ("get_receipt_example", "get_invoice_matching_example")
calls = []
native_send = httpx2.AsyncClient.send

def deny(*args, **kwargs):
    raise AssertionError("Non-MCP HTTP client blocked")

async def async_deny(*args, **kwargs):
    raise AssertionError("Non-MCP HTTP client blocked")

async def guarded_send(self, request, **kwargs):
    assert str(request.url) == ENDPOINT, "Only the exact public MCP endpoint is allowed"
    assert len(calls) < 20, "Request bound exceeded"
    assert not any(any(word in key.lower() for word in ("authorization", "payment", "api-key")) for key in request.headers), "Credential/payment header blocked"
    rpc = None
    if request.method == "POST":
        rpc = json.loads(await request.aread())
        assert rpc.get("method") in {"server/discover", "initialize", "notifications/initialized", "tools/list", "tools/call", "ping"}, "RPC blocked"
        assert "x402" not in json.dumps(rpc.get("params", {}).get("_meta", {})), "Payment metadata blocked"
        if rpc["method"] == "tools/call":
            assert rpc["params"]["name"] in FREE
            assert rpc["params"].get("arguments", {}) == {}
            assert not any(c.get("tool") == rpc["params"]["name"] for c in calls), "Only one call per example"
    else:
        assert request.method in {"GET", "DELETE"}
    entry = {"http_method": request.method, "rpc_method": rpc.get("method") if rpc else None,
             "tool": rpc["params"]["name"] if rpc and rpc["method"] == "tools/call" else None}
    calls.append(entry)
    kwargs["follow_redirects"] = False
    response = await native_send(self, request, **kwargs)
    entry["http_status"] = response.status_code
    return response

httpx2.AsyncClient.send = guarded_send
httpx2.Client.send = deny
httpx.AsyncClient.send = async_deny
httpx.Client.send = deny

from langchain.mcp import MCPAdapter

def canonical(value):
    return json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=False)

def verify_fixture(fixture):
    assert fixture["synthetic_only"] is True
    response = fixture.get("expected_response", fixture.get("response"))
    assert response["status"] == "review_required"
    expected = "res_" + hashlib.sha256(canonical(response["result"]).encode("utf-8")).hexdigest()
    assert response["identities"]["analysis_result_identity"] == expected

async def main():
    config = json.loads((ROOT / "langchain.connection.json").read_text(encoding="utf-8"))
    assert config == {"mcpServers": {"ardaro": {"url": ENDPOINT}}}
    report = {"framework": "LangChain MCPAdapter", "status": "FAIL", "endpoint": ENDPOINT,
              "started_at": datetime.now(timezone.utc).isoformat(), "python": platform.python_version(),
              "packages": {name: importlib.metadata.version(name) for name in ("langchain", "langchain-core", "fastmcp", "mcp", "httpx2")},
              "model_calls": 0, "paid_tool_calls": 0, "wallet_operations": 0, "cost_usdc": "0.00"}
    results = {}
    try:
        async with MCPAdapter(config) as adapter:
            tools = await adapter.list_tools()
            report["discovered_tool_names"] = [tool.name for tool in tools]
            report["discovered_tool_count"] = len(tools)
            report["server_info"] = adapter.client.server_info.model_dump(exclude_none=True)
            examples = {tool.name: tool for tool in tools if tool.name in FREE}
            assert set(examples) == set(FREE), "Exact named free tools were not exposed"
            for name in FREE:
                # A ToolCall obtains the native LangChain ToolMessage + structured artifact.
                message = await examples[name].ainvoke({"name": name, "args": {}, "id": name, "type": "tool_call"})
                assert message.status == "success"
                fixture = message.artifact["structured_content"]
                verify_fixture(fixture)
                results[name] = {"content": message.content, "structuredContent": fixture}
            report["free_examples_verified"] = list(FREE)
            report["status"] = "PASS"
    except BaseException as exc:
        report["error"] = {"name": type(exc).__name__, "message": str(exc)}
        raise
    finally:
        report["finished_at"] = datetime.now(timezone.utc).isoformat()
        report["calls"] = calls
        (ROOT / "execution-report.json").write_text(json.dumps(report, indent=2) + "\n", encoding="utf-8")
        (ROOT / "free-results.json").write_text(json.dumps(results, indent=2) + "\n", encoding="utf-8")
        print(json.dumps(report))

if __name__ == "__main__":
    asyncio.run(asyncio.wait_for(main(), timeout=90))

File: examples/frameworks/vercel/verify-free.mjs
// Actual AI SDK MCP adapter execution. The fetch boundary admits only this
// endpoint, lifecycle/discovery RPCs and two fixed examples with empty inputs.
import assert from 'node:assert/strict';
import {readFile,writeFile} from 'node:fs/promises';
import {createHash} from 'node:crypto';
import {createMCPClient} from '@ai-sdk/mcp';
import {z} from 'zod';
const endpoint='https://agents.getardaro.com/mcp';
const allowed=new Set(['get_receipt_example','get_invoice_matching_example']);
const calls=[],responses={};let discoveryCount=0,client;
const nativeFetch=globalThis.fetch;
const canonical=value=>JSON.stringify(value===null||typeof value!=='object'?value:Array.isArray(value)?value.map(v=>JSON.parse(canonical(v))):Object.fromEntries(Object.keys(value).sort().map(k=>[k,JSON.parse(canonical(value[k]))])));
const identity=value=>'res_'+createHash('sha256').update(canonical(value)).digest('hex');
async function guardedFetch(input,init={}) {
  const url=typeof input==='string'?input:input instanceof URL?input.href:input.url;
  assert.equal(url,endpoint,'No other endpoint allowed');
  assert.ok(calls.length<20,'Bounded request count exceeded');
  const headers=new Headers(init.headers??(input instanceof Request?input.headers:undefined));
  for(const key of headers.keys())assert.ok(!/authorization|payment|api-key/i.test(key),'Credential/payment header blocked');
  const method=init.method??(input instanceof Request?input.method:'GET');
  let rpc=null;
  if(method==='POST') {
    const body=typeof init.body==='string'?init.body:input instanceof Request?await input.clone().text():null;
    assert.equal(typeof body,'string');rpc=JSON.parse(body);
    assert.ok(['server/discover','initialize','notifications/initialized','tools/list','tools/call','ping'].includes(rpc.method),'Unapproved RPC');
    assert.ok(!JSON.stringify(rpc.params?._meta??{}).includes('x402'),'Payment metadata blocked');
    if(rpc.method==='tools/call'){assert.ok(allowed.has(rpc.params.name));assert.deepEqual(rpc.params.arguments??{},{});assert.ok(!calls.some(c=>c.tool===rpc.params.name),'Only one call per free fixture');}
  } else assert.ok(['GET','DELETE'].includes(method),'Unapproved transport method');
  const entry={http_method:method,rpc_method:rpc?.method??null,tool:rpc?.method==='tools/call'?rpc.params.name:null};calls.push(entry);
  const response=await nativeFetch(input,{...init,redirect:'error',signal:AbortSignal.any([...(init.signal?[init.signal]:[]),AbortSignal.timeout(20000)])});entry.http_status=response.status;
  if(rpc?.method==='tools/list'&&response.ok){const raw=await response.clone().json();discoveryCount=raw.result?.tools?.length??0;}
  return response;
}
const config=JSON.parse(await readFile(new URL('./vercel.connection.json',import.meta.url),'utf8'));
assert.deepEqual(config,{transport:{type:'http',url:endpoint}});
const report={framework:'Vercel AI SDK MCP',status:'FAIL',endpoint,started_at:new Date().toISOString(),node:process.version,packages:{'@ai-sdk/mcp':JSON.parse(await readFile(new URL('./node_modules/@ai-sdk/mcp/package.json',import.meta.url),'utf8')).version,zod:JSON.parse(await readFile(new URL('./node_modules/zod/package.json',import.meta.url),'utf8')).version},model_calls:0,paid_tool_calls:0,wallet_operations:0,cost_usdc:'0.00'};
try {
  client=await createMCPClient({...config,transport:{...config.transport,fetch:guardedFetch},maxRetries:0});
  report.server_info=client.serverInfo;report.protocol_version=client.initializeResult.protocolVersion;
  const tools=await client.tools({schemas:{get_receipt_example:{inputSchema:z.object({}).strict()},get_invoice_matching_example:{inputSchema:z.object({}).strict()}}});
  assert.deepEqual(Object.keys(tools).sort(),[...allowed].sort());
  for(const name of allowed){const result=await tools[name].execute({}, {toolCallId:name,messages:[]});assert.notEqual(result.isError,true);const fixture=result.structuredContent;assert.equal(fixture?.synthetic_only,true);const response=fixture.expected_response??fixture.response;assert.equal(response?.status,'review_required');assert.equal(response.identities.analysis_result_identity,identity(response.result));responses[name]=result;}
  report.status='PASS';report.discovered_tool_count=discoveryCount;report.free_examples_verified=[...allowed];
} catch(error){report.error={name:error.name,message:error.message};process.exitCode=1;}
finally {if(client)await client.close();report.finished_at=new Date().toISOString();report.calls=calls;await writeFile('execution-report.json',JSON.stringify(report,null,2)+'\n','utf8');await writeFile('free-results.json',JSON.stringify(responses,null,2)+'\n','utf8');console.log(JSON.stringify(report));}
